Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingFeatureProcessor.cpp

```diff
@@ -85,6 +86,157 @@ namespace AZ
+            }
+
+            geometryTypeHandle->m_bindlessBufferIndex = bindlessBufferIndex;
+            m_proceduralGeometryInfoBufferNeedsUpdate = true;
+        }
+
+        void RayTracingFeatureProcessor::AddProceduralGeometry(
+            ProceduralGeometryTypeWeakHandle geometryTypeHandle,
+            const Uuid& uuid,
+            const Aabb& aabb,
+            const SubMeshMaterial& material,
+            RHI::RayTracingAccelerationStructureInstanceInclusionMask instanceMask,
+            uint32_t localInstanceIndex)
+        {
+            if (!m_rayTracingEnabled)
+            {
+                return;
+            }
+
+            AZStd::unique_lock<AZStd::mutex> lock(m_mutex);
+
+            RHI::Ptr<RHI::RayTracingBlas> rayTracingBlas = AZ::RHI::RayTracingBlas::CreateRHIRayTracingBlas();
+            RHI::Ptr<RHI::Device> device = RHI::RHISystemInterface::Get()->GetDevice();
+            RHI::RayTracingBlasDescriptor blasDescriptor;
+            blasDescriptor.Build()
+                ->AABB(aabb)
+                ;
+            rayTracingBlas->CreateBuffers(*device, &blasDescriptor, *m_bufferPools);
+
+            ProceduralGeometry proceduralGeometry;
+            proceduralGeometry.m_uuid = uuid;
+            proceduralGeometry.m_typeHandle = geometryTypeHandle;
+            proceduralGeometry.m_aabb = aabb;
+            proceduralGeometry.m_instanceMask = static_cast<uint32_t>(instanceMask);
+            proceduralGeometry.m_blas = rayTracingBlas;
+            proceduralGeometry.m_localInstanceIndex = localInstanceIndex;
+            m_proceduralGeometryLookup.emplace(uuid, m_proceduralGeometry.size());
+            m_proceduralGeometry.push_back(proceduralGeometry);
+
+            geometryTypeHandle->m_instanceCount++;
+
+            MaterialInfo& materialInfo = m_proceduralGeometryMaterialInfos.emplace_back();
+            ConvertMaterial(materialInfo, material);
+
+            MeshBlasInstance meshBlasInstance;
+            meshBlasInstance.m_count = 1;
+            meshBlasInstance.m_subMeshes.push_back(SubMeshBlasInstance{ rayTracingBlas });
+            m_blasInstanceMap.emplace(Data::AssetId(uuid), meshBlasInstance);
+
+            m_revision++;
+            m_proceduralGeometryInfoBufferNeedsUpdate = true;
+            m_materialInfoBufferNeedsUpdate = true;
+            m_indexListNeedsUpdate = true;
+        }
+
+        void RayTracingFeatureProcessor::SetProceduralGeometryTransform(
+            const Uuid& uuid, const Transform& transform, const Vector3& nonUniformScale)
+        {
+            if (!m_rayTracingEnabled)
+            {
+                return;
+            }
+
+            if (auto it = m_proceduralGeometryLookup.find(uuid); it != m_proceduralGeometryLookup.end())
+            {
+                m_proceduralGeometry[it->second].m_transform = transform;
+                m_proceduralGeometry[it->second].m_nonUniformScale = nonUniformScale;
+            }
+
+            m_revision++;
```

Comment:
In current engine code at least `AddMesh` and `RemoveMesh` are likely to be called concurrently, as multiple jobs for adding/removing RT data are used in `MeshFeatureProcessor::Simulate`. Some of the other functions are called from non-concurrent functions (eg. `SetupFrameGraphDependencies`, `OnRenderEnd`), others from events (eg. `OnTransformChanged`). For procedural geometry, eg. the sphere the creation is in `EditorDebugDrawSphereComponent::Activate`, and the transform is updated in `OnBeginPrepareRender`.
So would it make sense to just lock every access to FP data since it cannot be guaranteed that some future code changes access the functions concurrently (not sure if this is already the case with Activate/OnTransformChanged and the like).